test/test-addon.js

```diff
@@ -26,3 +42,33 @@ test('build simple addon', function (t) {
   proc.stdout.setEncoding('utf-8')
   proc.stderr.setEncoding('utf-8')
 })
+
+test('build simple addon - teardown', cleanup(path.join(addonPath, 'build')))
+
+test('build with different build dir - setup', cleanup(path.join(addonPath, 'foo')))
+
+test('build with different build dir', function(t) {
+  var cmd = [nodeGyp, 'rebuild', '-C',
+              addonPath, '--loglevel=verbose', '--build-dir=foo']
+  var proc = execFile(process.execPath, cmd, function (err, stdout, stderr) {
+    var logLines = stderr.toString().trim().split(/\r?\n/)
+    var lastLine = logLines[logLines.length-1]
+    t.strictEqual(err, null)
+    t.strictEqual(lastLine, 'gyp info ok', 'should end in ok')
+
+    try {
+      var binding = require(path.join(addonPath, 'foo/Release/hello.node'))
+      t.strictEqual(binding.hello(), 'world')
+
+      // Cleanup
+      exec('node-gyp --build-dir=foo clean', t.end)
```

Comment:
Can you use `execFile(process.execPath, [nodeGyp, ...])` here?  Calling out to node-gyp risks picking up the version that is on the path, not what's under test.

Aside: I know they're Rod's work but can you wrap lines at 80 columns?  Thanks. :-)